services/document-manager/backend/app/docman/resources/document_version_list_resource.py

```diff
@@ -0,0 +1,68 @@
+import base64
+import time
+import uuid
+from datetime import datetime
+from urllib.parse import urlparse
+from wsgiref.handlers import format_date_time
+
+from app.docman.utils.document_upload_helper import DocumentUploadHelper
+import requests
+from app.docman.models.document_version import DocumentVersion
+from app.config import Config
+from app.constants import OBJECT_STORE_PATH, OBJECT_STORE_UPLOAD_RESOURCE, FILE_UPLOAD_SIZE, FILE_UPLOAD_OFFSET, \
+    FILE_UPLOAD_PATH, FILE_UPLOAD_EXPIRY, TIMEOUT_24_HOURS, TUS_API_VERSION, TUS_API_SUPPORTED_VERSIONS, \
+    FORBIDDEN_FILETYPES
+from app.docman.models.document import Document
+from app.extensions import api, cache
+from app.utils.access_decorators import requires_any_of, DOCUMENT_UPLOAD_ROLES
+from flask import request, current_app, make_response, jsonify
+from flask_restplus import Resource, reqparse
+from werkzeug.exceptions import BadRequest, NotFound, RequestEntityTooLarge, InternalServerError, BadGateway
+
+CACHE_TIMEOUT = TIMEOUT_24_HOURS
+
+
+@api.route(f'/documents/<string:document_guid>/versions')
+class DocumentVersionListResource(Resource):
+    parser = reqparse.RequestParser(trim=True)
+    parser.add_argument(
+        'filename', type=str, required=False, help='File name + extension of the document.')
+
+    @requires_any_of(DOCUMENT_UPLOAD_ROLES)
+    def post(self, document_guid):
+        """
+        Initiating a new Tusd document upload to the storage path of the 
+        original document, and creates a new DocumentVersion to 
+        keep a record of the new document version
+        """
+        document = Document.query.filter_by(
+            document_guid=document_guid).one_or_none()
+
+        if not document:
+            raise NotFound('Document not found')
+
+        file_size, data, filename = DocumentUploadHelper().parse_and_validate_uploaded_file(
+            self.parser.parse_args())
+
+        document_guid = str(document.document_guid)
+        version_guid = str(uuid.uuid4())
+
+        response, object_store_path = DocumentUploadHelper.initiate_document_upload(
+            document_guid=document_guid,
+            file_path=document.full_storage_path,
+            folder=None,
+            file_size=file_size,
+            version_guid=version_guid)
+
+        # Create document record
+        new_version = DocumentVersion(
+            id=version_guid,
+            document_guid=document.document_guid,
+            created_by='mds',
+            created_date=datetime.utcnow(),
+            file_display_name=filename,
+            upload_started_date=datetime.utcnow(),
```

Comment:
I'm probably missing something, but:
1. Will created_date and upload_started_date always be the same values? Do we _need_ both?
2. Is there a different spot where new versions uploaded by users are processed where their username gets put in the created_by field?
3. The bit with the id/version_guid: on ln 48 a version_guid is generated, but ln 58 it could just as easily be handled by the db, since it doesn't use the response from ln 50. Was the object_store_path meant to be passed in to the DocumentVersion?
4. Nitpick: I don't like the column names where "created_date" is a _timestamp_.

Reply:
Answers to 1., and 4.: I do not like it particularly either. In practice created_date and upload_started_date will be the same (can't think of a scenario they wouldn't). There might be a legacy reason for keeping them both on the Document modal that I'm not aware of (maybe import of documents from MMS?). The only reason I kept both is to keep it as consistent with the `Document` model as I could. Happy to change it up though

Reply:
The reason for generating the version_guid "manually" is because the version is passed along to Tusd to initialize the document upload (document_upload_helper line 22), that way we will avoid orphan `DocumentVersions` if it gets created by the DB and the Tusd initialization fails for any reason.